Log when publishing Ivy or Maven metadata with known incompatible features from Gradle

ljacomet opened the pull request:
Features like constraints, platforms or declared capabilities are not
supported in the ivy.xml file format.

Features like dynamic versions, with `latest.<status>` or `1.+`, and capabilities are not
supported in the pom.xml file format.

ljacomet commented:
I have marked this as WIP because it needs further testing and possible coverage of other cases.
However I am interested in feedback on the approach, the perceived value for user and the logging of that information at `LIFECYCLE` level.

I had a look at using `SingleMessageLogger` but the presence of the stacktrace makes it really scary IMO. And similarly for a `WARN` level logging.
And if as a user you do not care, due to your usage pattern - pom consumed by Gradle only for example - these notifications are just noise.

ljacomet commented:
This is complete given its intended scope.

For `ivy-publish`, will log when any of the following is declared:
* Capabilities
* Constraints
* Platforms
* Dependencies without version

For `maven-publish`, will log when any of the following is declared:
* Capabilities
* Dependencies using `+` in the version
* Dependencies using `latest.<level>` where `level` is neither `release` nor `integration`

bigdaz commented on subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publication/DefaultMavenPublication.java:
```diff
@@ -267,7 +283,38 @@ private void populateFromComponent() {
+                    unsupportedUsages.add(String.format("Declares capability %s:%s:%s", capability.getGroup(), capability.getName(), capability.getVersion()));
+                }
+            }
+        }
```
I think we are missing validation for dependencies declared with attributes.

ljacomet replied:
Added

bigdaz commented on subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/publication/DefaultIvyPublication.java:
```diff
@@ -245,15 +257,54 @@ private void populateFromComponent() {
+            }
+            formatter.endChildren();
+            LOG.lifecycle(formatter.toString());
+        }
```
Most of this validation code can be extracted into a utility that is shared between Maven and Ivy publishing. This might be a simple builder that allows you to add an 'unsupportedUsage', and it constructs the message.

ljacomet replied:
Done

bigdaz commented on subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/publication/DefaultIvyPublication.java:
```diff
@@ -94,6 +103,8 @@ public int compare(UsageContext left, UsageContext right) {
         }
     };
+    @VisibleForTesting
+    public static final String INCOMPATIBLE_FEATURE = " uses dependency management feature(s) that will most likely make the produced Ivy file incomplete and/or not consumable by Ivy compatible solutions";
```
This message is a bit wordy for my liking:

"contains dependencies that cannot be represented in a published [pom file | ivy descriptor]."
"contains dependencies that will produce a pom file that cannot be consumed by a Maven client"
"contains dependencies that will produce an Ivy descriptor that cannot be consumed by an Ivy client"

I'm not sure that there's a real use case for the last one.

ljacomet replied:
Reworked based on your suggestion

ljacomet commented:
@bigdaz I have given this a go and addressed your remarks aside from having a user control to silence some warnings. Please have another look as this could still make it into 5.2
